polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -419,21 +421,58 @@ impl<H: Copy> From<CandidateReceiptV2<H>> for super::v8::CandidateReceipt<H> {
 
 /// A strictly increasing sequence number, typically this would be the least significant byte of the
 /// block number.
-#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, TypeInfo, RuntimeDebug)]
 pub struct CoreSelector(pub u8);
 
 /// An offset in the relay chain claim queue.
-#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, TypeInfo, RuntimeDebug)]
 pub struct ClaimQueueOffset(pub u8);
 
+/// Approved PeerId type. PeerIds in polkadot should typically be 32 bytes long but for identity
+/// multihash can go up to 64. Cannot reuse the PeerId type definition from the networking code as
+/// it's too generic and extensible.
+pub type ApprovedPeerId = BoundedVec<u8, ConstU32<64>>;
+
 /// Signals that a parachain can send to the relay chain via the UMP queue.
 #[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
 pub enum UMPSignal {
 	/// A message sent by a parachain to select the core the candidate is committed to.
 	/// Relay chain validators, in particular backers, use the `CoreSelector` and
 	/// `ClaimQueueOffset` to compute the index of the core the candidate has committed to.
 	SelectCore(CoreSelector, ClaimQueueOffset),
+	/// A message sent by a parachain to promote the reputation of a given peerid.
+	ApprovedPeer(ApprovedPeerId),
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug, Default)]
+/// User-friendly representation of a candidate's UMP signals.
+pub struct CandidateUMPSignals {
+	select_core: Option<(CoreSelector, ClaimQueueOffset)>,
+	approved_peer: Option<ApprovedPeerId>,
 }
+
+impl CandidateUMPSignals {
+	/// Get the core selector UMP signal.
+	pub fn core_selector(&self) -> Option<(CoreSelector, ClaimQueueOffset)> {
+		self.select_core
+	}
+
+	/// Get a reference to the approved peer UMP signal.
+	pub fn approved_peer(&self) -> Option<&ApprovedPeerId> {
```

Comment:
Feels inconsistent to return a ref on this on and, core selector does not.

Reply:
This is because core selector components implement copy (they're just 2 bytes). The user of the code can choose whether they want to clone the peerid (which is larger, 64 bytes)